src/err/mod.rs

```diff
@@ -327,15 +333,70 @@ impl PyErr {
             Some(obj) => obj.as_ptr(),
         };
 
-        unsafe {
-            let null_terminated_name =
-                CString::new(name).expect("Failed to initialize nul terminated exception name");
+        let null_terminated_name =
+            CString::new(name).expect("Failed to initialize nul terminated exception name");
+
+        let ptr = unsafe {
+            ffi::PyErr_NewException(null_terminated_name.as_ptr(), base, dict)
+                as *mut ffi::PyTypeObject
+        };
+        match NonNull::new(ptr) {
+            Some(not_null) => not_null,
+            None => panic!("{}", PyErr::fetch(py)),
+        }
+    }
 
-            NonNull::new_unchecked(ffi::PyErr_NewException(
-                null_terminated_name.as_ptr() as *mut c_char,
+    /// Creates a new exception type with the given name and docstring.
+    ///
+    /// - `base` can be an existing exception type to subclass, or a tuple of classes.
+    /// - `dict` specifies an optional dictionary of class variables and methods.
+    /// - `doc` will be the docstring seen by python users.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if:
+    /// - `name` is not of the form `<module>.<ExceptionName>`.
+    /// - `name` or `doc` cannot be converted to [`CString`]s.
+    pub fn new_type_with_doc<'p>(
```

Comment:
We could also just merge this and new_doc, if a breaking change is acceptable here.

Reply:
Hmmm. I'm really on the fence about this. I do think that a `_with_doc` method should probably take `&str` rather than `Option<&str>`. These two methods can probably share a common implementation which takes option.

Otoh, I feel more strongly that the return value of `new_type` should be changed (see other comment). So that does open up the chance for us to make more breakages.

Which design do you prefer? Single method taking option or two methods?

Reply:
I'd definitely prefer a single method that takes an option.